R/ngrams.R

```diff
@@ -0,0 +1,117 @@
+#' Get ngrams from a corpus
+#'
+#' @param data Data frame of documents containing a text column
+#' @param n The number of words to include in the token (e.g., n = 2 for bigram)
+#' @param text_col The name of the column where the text to be tokenised is located
+#' @param title_col The name of the title column
+#' @param stem If `TRUE` create additional columns with stemmed version of tokens
+#' @param lemma If `TRUE` create additional columns with lemmatised version of tokens
+#'
+#' @return A data frame with an "ngram" column and columns for its constituent words
+#' @export
+get_ngrams <- function(data,
+                       n,
+                       text_col = NULL,
+                       title_col = NULL,
+                       rm_n_short_words = 2,
+                       min_freq = 5,
+                       stem = FALSE,
+                       lemma = FALSE) {
+
+  # Create ngrams column and separate columns for each of the n terms
+  data_ngrams <- data |>
+    dplyr::select(title_col, text_col) |>
+    tidytext::unnest_tokens(ngram, text_col, token = "ngrams", n = n) |>
+    tidyr::separate(ngram, into = paste("w" , 1:n, sep = "_"),
+                    sep = " ", remove = FALSE)
+
+  # Remove ngrams below a given frequency
+  ngram_counts <- data_ngrams |>
+    count(ngram, sort = TRUE) |>
+    filter(n > min_freq)
+
+  data_ngrams <- data_ngrams |>
+    filter(ngram %in% ngram_counts$ngram)
+
+  # Remove stop words from each column starting with "w_"
+  w_cols <- names(data_ngrams)[grepl("w_", names(data_ngrams))]
+
+  # TODO short words in a separate function
+  short_words <- data |>
+    select(text_raw) |>
+    unnest_tokens(output = word, input = text_raw) |>
+    anti_join(tidytext::stop_words, by = "word") |>  # remove stop words
+    mutate(word = textstem::lemmatize_words(word)) |>  # lemmatise words
+    filter(nchar(word) <= rm_n_short_words)
+
+  for (i in w_cols) {
+    data_ngrams <- data_ngrams |>
+      anti_join(stop_words, by = set_names("word", i)) |>
+      anti_join(short_words, by = set_names("word", i)) |>
+      anti_join(urbanism_stopwords(add_stopwords = c("emu", "tu", "delft", "ku",
+                                                     "leuven", "upc", "barcelona",
+                                                     "iuav", "venice"),
+                                   convert_to_regex = FALSE) %>%
+                  data.frame(word = .),
+                by = set_names("word", i))
+  }
+
+  # Add columns with stemmed versions of individual terms if `stem = TRUE`
+  if (stem) {
+    data_ngrams <- data_ngrams |>
+      dplyr::mutate(dplyr::across(
+        tidyselect::starts_with("w_"),
```

Comment:
As you've already have the vector with the names of the words columns `w_cols`, you can use it here directly

```suggestion
        w_cols,
```